File: android/versioned-abis/expoview-abi30_0_0/src/main/java/abi30_0_0/host/exp/exponent/modules/api/screens/ScreenContainerViewManager.java
package abi30_0_0.host.exp.exponent.modules.api.screens;

import android.view.View;

import abi30_0_0.com.facebook.react.module.annotations.ReactModule;
import abi30_0_0.com.facebook.react.uimanager.ThemedReactContext;
import abi30_0_0.com.facebook.react.uimanager.ViewGroupManager;

@ReactModule(name = ScreenContainerViewManager.REACT_CLASS)
public class ScreenContainerViewManager extends ViewGroupManager<ScreenContainer> {

  protected static final String REACT_CLASS = "RNSScreenContainer";

  @Override
  public String getName() {
    return REACT_CLASS;
  }

  @Override
  protected ScreenContainer createViewInstance(ThemedReactContext reactContext) {
    return new ScreenContainer(reactContext);
  }

  @Override
  public void addView(ScreenContainer parent, View child, int index) {
    if (!(child instanceof Screen)) {
      throw new IllegalArgumentException("Attempt attach child that is not of type RNScreen");
    }
    parent.addScreen((Screen) child, index);
  }

  @Override
  public void removeViewAt(ScreenContainer parent, int index) {
    parent.removeScreenAt(index);
  }

  @Override
  public int getChildCount(ScreenContainer parent) {
    return parent.getScreenCount();
  }

  @Override
  public View getChildAt(ScreenContainer parent, int index) {
    return parent.getScreenAt(index);
  }
}

File: android/vendored/unversioned/react-native-screens/android/src/main/java/com/swmansion/rnscreens/ScreenContainerViewManager.kt
package com.swmansion.rnscreens

import android.view.View
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.LayoutShadowNode
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager

@ReactModule(name = ScreenContainerViewManager.REACT_CLASS)
class ScreenContainerViewManager : ViewGroupManager<ScreenContainer<*>>() {
    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(reactContext: ThemedReactContext): ScreenContainer<ScreenFragment> = ScreenContainer(reactContext)

    override fun addView(parent: ScreenContainer<*>, child: View, index: Int) {
        require(child is Screen) { "Attempt attach child that is not of type RNScreens" }
        parent.addScreen(child, index)
    }

    override fun removeViewAt(parent: ScreenContainer<*>, index: Int) {
        parent.removeScreenAt(index)
    }

    override fun removeAllViews(parent: ScreenContainer<*>) {
        parent.removeAllScreens()
    }

    override fun getChildCount(parent: ScreenContainer<*>): Int = parent.screenCount

    override fun getChildAt(parent: ScreenContainer<*>, index: Int): View = parent.getScreenAt(index)

    override fun createShadowNodeInstance(context: ReactApplicationContext): LayoutShadowNode = ScreensShadowNode(context)

    override fun needsCustomLayoutForChildren(): Boolean = true

    companion object {
        const val REACT_CLASS = "RNSScreenContainer"
    }
}
